desc/integrals/_bounce_utils.py

```diff
@@ -700,127 +845,315 @@ def interp_to_argmin_hard(h, points, knots, g, dg_dz, method="cubic"):
     # to accumulate through only the minimum since we are differentiating how our
     # physics objective changes wrt equilibrium perturbations not wrt which of the
     # extrema get interpolated to.
     argmin = jnp.argmin(
-        _where_for_argmin(z1, z2, ext, g_ext, jnp.max(g_ext) + 1),
+        _where_for_argmin(points, ext, g_ext, jnp.max(g_ext) + 1),
         axis=-1,
     )
-    h = interp1d_vec(
+    return interp1d_vec(
         jnp.take_along_axis(ext[jnp.newaxis], argmin, axis=-1),
         knots,
         h[..., jnp.newaxis, :],
         method=method,
     )
-    assert h.shape == z1.shape
-    return h
 
 
-def plot_ppoly(
-    ppoly,
-    num=1000,
-    z1=None,
-    z2=None,
-    k=None,
-    k_transparency=0.5,
-    klabel=r"$k$",
-    title=r"Intersects $z$ in epigraph($f$) s.t. $f(z) = k$",
-    hlabel=r"$z$",
-    vlabel=r"$f$",
-    show=True,
-    start=None,
-    stop=None,
-    include_knots=False,
-    knot_transparency=0.2,
-    include_legend=True,
+def interp_fft_to_argmin(
+    NFP, T, h, points, knots, g, dg_dz, beta=-100, upper_sentinel=1e2
 ):
-    """Plot the piecewise polynomial ``ppoly``.
+    """Interpolate ``h`` to the deepest point of ``g`` between ``z1`` and ``z2``.
+
+    Let E = {ζ ∣ ζ₁ < ζ < ζ₂} and A = argmin_E g(ζ). Returns mean_A h(ζ).
 
     Parameters
     ----------
-    ppoly : PPoly
-        Piecewise polynomial f.
-    num : int
-        Number of points to evaluate for plot.
-    z1 : jnp.ndarray
-        Shape (k.shape[0], W).
-        Optional, intersects with ∂f/∂z <= 0.
-    z2 : jnp.ndarray
-        Shape (k.shape[0], W).
-        Optional, intersects with ∂f/∂z >= 0.
-    k : jnp.ndarray
-        Shape (k.shape[0], ).
-        Optional, k such that f(z) = k.
-    k_transparency : float
-        Transparency of intersect lines.
-    klabel : str
-        Label of intersect lines.
-    title : str
-        Plot title.
-    hlabel : str
-        Horizontal axis label.
-    vlabel : str
-        Vertical axis label.
-    show : bool
-        Whether to show the plot. Default is true.
-    start : float
-        Minimum z on plot.
-    stop : float
-        Maximum z on plot.
-    include_knots : bool
-        Whether to plot vertical lines at the knots.
-    knot_transparency : float
-        Transparency of knot lines.
-    include_legend : bool
-        Whether to include the legend in the plot. Default is true.
+    T : PiecewiseChebyshevSeries
+        Set of 1D Chebyshev spectral coefficients of θ along field line.
+        {θ_α : ζ ↦ θ(α, ζ) | α ∈ A} where A = (α₀, α₁, …, αₘ₋₁) is the same
+        field line.
+    h : jnp.ndarray
+        Shape (..., grid.num_theta, grid.num_zeta)
+        Periodic function evaluated on tensor-product grid in (ρ, θ, ζ) with
+        uniformly spaced nodes (θ, ζ) ∈ [0, 2π) × [0, 2π/NFP).
+        Preferably power of 2 for ``grid.num_theta`` and ``grid.num_zeta``.
+    points : jnp.ndarray
+        Shape (..., num well).
```

Comment:
Can you explain the type hint here? what should the shape of points be?

Reply:
usually it's whatever is returned by `bounce.points`. The type hint is stating that the function only assumes that the last axis enumerates the wells. it can be 1d dimensional or ndimensional and it would still work fine with standard numpy broadcasting rules.